pkg/standalone-utils/contracts/PermitRelayer.sol

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-vault/contracts/interfaces/IVault.sol";
+import "./base/Multicall.sol";
+import "./base/VaultPermit.sol";
+import "./interfaces/IPermit.sol";
+import "./interfaces/IPermitDAI.sol";
+
+/**
+ * @title PermitRelayer
+ * @notice Allows users to atomically call permit on multiple tokens and then interact with the Balancer Vault
+ * @dev This relayer is extremely powerful and has full control over users' assets should they authorise it.
+ * For this reason, users should only perform signature-based approvals to allow it to perform only a single action.
+ *
+ * WARNING: Do not call setRelayerApproval on the Vault for this relayer - doing so WILL result in loss of funds.
```

Comment:
We're always spending `msg.sender`'s funds, why would it be an issue for a user to grant approval to this contract?

Reply:
It's not. I wrote this line before I added more checks to restrict `funds.sender`.

Reply:
In that case, wouldn't it make sense to let users use such a relayer _without_ having to provide signatures? This otherwise has you sign n messages, plus the final transaction.

Reply:
(as an unrelated comment, we shouldn't ever have governance approved contracts that are not safe for a user to approve, that is, that must only be used via signatures. This would be difficult to communicate, and we shouldn't have hard to understand _and_ dangerous function calls available.

If such a use case existed (like we had here before that change), then what the relayer should do is `require(!vault.hasApprovedRelayer(user, address(this)))` to be absolutely sure that any successful transactions happen via signatures)

Reply:
Sorry, I'm agreeing with you. This line is a holdover from when things were less secure so that would have been an issue.

Reply:
We should add another function to allow setting an approval atomically with the first signature based interaction.

Reply:
Ah I must not have expressed myself correctly, I did get that you agreed with me :sweat_smile: My two comments above mean:
 1) would it make sense to have a multicall relayer that doesn't use signatures, for simpler UX?
 2) we should never make a relayer that is unsafe to approve by a user